flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java

```diff
@@ -0,0 +1,703 @@
+        }
+
+        if (jobState != null) {
+            deploymentStatus.getJobStatus().setState(jobState);
+        }
+
+        deploymentStatus.setLastReconciledTimestamp(System.currentTimeMillis());
+        flinkBlueGreenDeployment.setStatus(deploymentStatus);
+        return UpdateControl.patchStatus(flinkBlueGreenDeployment);
+    }
+
+    private void deploy(
+            FlinkBlueGreenDeployment bgDeployment,
+            DeploymentType deploymentType,
+            Savepoint lastCheckpoint,
+            Context<FlinkBlueGreenDeployment> josdkContext,
+            boolean isFirstDeployment)
+            throws JsonProcessingException {
+        ObjectMeta bgMeta = bgDeployment.getMetadata();
+
+        // Deployment
+        FlinkDeployment flinkDeployment = new FlinkDeployment();
+        flinkDeployment.setApiVersion("flink.apache.org/v1beta1");
+        flinkDeployment.setKind("FlinkDeployment");
+        FlinkBlueGreenDeploymentSpec spec = bgDeployment.getSpec();
+
+        String childDeploymentName =
+                bgMeta.getName() + "-" + deploymentType.toString().toLowerCase();
+
+        FlinkBlueGreenDeploymentSpec adjustedSpec =
+                adjustNameReferences(
+                        spec,
+                        bgMeta.getName(),
+                        childDeploymentName,
+                        "spec",
+                        FlinkBlueGreenDeploymentSpec.class);
+
+        if (lastCheckpoint != null) {
+            String location = lastCheckpoint.getLocation().replace("file:", "");
+            LOG.info("Using checkpoint: " + location);
+            adjustedSpec.getTemplate().getSpec().getJob().setInitialSavepointPath(location);
+        }
+
+        flinkDeployment.setSpec(adjustedSpec.getTemplate().getSpec());
+
+        // Deployment metadata
+        ObjectMeta flinkDeploymentMeta = getDependentObjectMeta(bgDeployment);
+        flinkDeploymentMeta.setName(childDeploymentName);
+        flinkDeploymentMeta.setLabels(
+                Map.of(deploymentType.getClass().getSimpleName(), deploymentType.toString()));
+        flinkDeployment.setMetadata(flinkDeploymentMeta);
+
+        // Deploy
+        josdkContext.getClient().resource(flinkDeployment).createOrReplace();
+    }
+
+    private static void deleteDeployment(
+            FlinkDeployment currentDeployment, Context<FlinkBlueGreenDeployment> josdkContext) {
+        String deploymentName = currentDeployment.getMetadata().getName();
+        List<StatusDetails> deletedStatus =
+                josdkContext
+                        .getClient()
+                        .resources(FlinkDeployment.class)
+                        .inNamespace(currentDeployment.getMetadata().getNamespace())
+                        .withName(deploymentName)
+                        .delete();
+
+        boolean deleted =
+                deletedStatus.size() == 1
+                        && deletedStatus.get(0).getKind().equals("FlinkDeployment");
+
```

Comment:
do we want to wait here until the deletion is actually finished (the object is not there anymore?) that would require some additional logic here